controlplane/kubeadm/internal/kubeadm_config_map.go

```diff
@@ -196,10 +197,68 @@ func (k *kubeadmConfig) UpdateCoreDNSImageInfo(repository, tag string) error {
 	return nil
 }
 
+// UpdateAPIServer sets the api server configuration to values set in `apiServer` in kubeadm config map.
+func (k *kubeadmConfig) UpdateAPIServer(apiServer kubeadmv1.APIServer) (bool, error) {
+	data, ok := k.ConfigMap.Data[clusterConfigurationKey]
+	if !ok {
+		return false, errors.Errorf("unable to find %q in kubeadm ConfigMap", clusterConfigurationKey)
+	}
+
+	configuration, err := yamlToUnstructured([]byte(data))
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to decode kubeadm ConfigMap's %q to Unstructured object", clusterConfigurationKey)
+	}
+
+	apiServerPath := []string{"apiServer"}
+	currentAPIServerConfig, _, err := unstructured.NestedFieldCopy(configuration.UnstructuredContent(), apiServerPath...)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to retrieve %q from kubeadm ConfigMap", strings.Join(apiServerPath, "."))
+	}
+
+	// convert APIServer to map[string]interface because unstructured.SetNestedField does not accept custom structs.
+	newAPIServerConfig, err := toMap(apiServer)
+	if err != nil {
+		return false, errors.Wrap(err, "Unable to convert api server to map")
+	}
+
+	// if there are no changes, return early.
+	if reflect.DeepEqual(newAPIServerConfig, currentAPIServerConfig) {
+		return false, nil
+	}
+
+	if err := unstructured.SetNestedField(configuration.UnstructuredContent(), newAPIServerConfig, apiServerPath...); err != nil {
+		return false, errors.Wrapf(err, "unable to update %q on kubeadm ConfigMap", strings.Join(apiServerPath, "."))
+	}
+
+	updated, err := yaml.Marshal(configuration)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to encode kubeadm ConfigMap's %q to YAML", clusterConfigurationKey)
+	}
+
+	k.ConfigMap.Data[clusterConfigurationKey] = string(updated)
+	return true, nil
+}
+
 // yamlToUnstructured looks inside a config map for a specific key and extracts the embedded YAML into an
 // *unstructured.Unstructured.
 func yamlToUnstructured(rawYAML []byte) (*unstructured.Unstructured, error) {
 	unst := &unstructured.Unstructured{}
 	err := yaml.Unmarshal(rawYAML, unst)
 	return unst, err
 }
+
+// toMap converts obj of a concrete type to map[string]interface{}.
+// This is useful for setting values in unstructured objects which expects nested fields to be of type map[string]interface.
+func toMap(obj interface{}) (map[string]interface{}, error) {
```

Comment:
The implementation is a lot more complex underneath than what you have here, but if you don't want to write this method yourself, you can use apimachinery's `runtime.DefaultUnstructuredConverter.ToUnstructured()`

Reply:
Thanks for letting me know this!